crates/core_arch/src/riscv_shared/zk.rs

```diff
@@ -0,0 +1,594 @@
+///
+/// # Safety
+///
+/// This function is safe to use if the `zbkb` target feature is present.
+#[target_feature(enable = "zbkb")]
+#[cfg_attr(test, assert_instr(pack))]
+#[inline]
+pub unsafe fn pack(rs1: usize, rs2: usize) -> usize {
+    let value: usize;
+    unsafe {
+        asm!(
+            "pack {rd},{rs1},{rs2}",
+            rd = lateout(reg) value,
+            rs1 = in(reg) rs1,
+            rs2 = in(reg) rs2,
+            options(pure, nomem, nostack),
+        )
+    }
+    value
+}
+
+/// Pack the low bytes of rs1 and rs2 into rd.
+///
+/// And the packh instruction packs the least-significant bytes of rs1 and rs2 into the 16
+/// least-significant bits of rd, zero extending the rest of rd.
+///
+/// Source: RISC-V Cryptography Extensions Volume I: Scalar & Entropy Source Instructions
+///
+/// Version: v1.0.1
+///
+/// Section: 3.18
+///
+/// # Safety
+///
+/// This function is safe to use if the `zbkb` target feature is present.
+#[target_feature(enable = "zbkb")]
+#[cfg_attr(test, assert_instr(packh))]
+#[inline]
+pub unsafe fn packh(rs1: usize, rs2: usize) -> usize {
+    let value: usize;
+    unsafe {
+        asm!(
+            "packh {rd},{rs1},{rs2}",
+            rd = lateout(reg) value,
+            rs1 = in(reg) rs1,
+            rs2 = in(reg) rs2,
+            options(pure, nomem, nostack),
+        )
+    }
+    value
+}
+
+/// Reverse the bits in each byte of a source register.
+///
+/// This instruction reverses the order of the bits in every byte of a register.
+///
+/// Source: RISC-V Cryptography Extensions Volume I: Scalar & Entropy Source Instructions
+///
+/// Version: v1.0.1
+///
+/// Section: 3.13
+///
+/// # Safety
+///
+/// This function is safe to use if the `zbkb` target feature is present.
+#[target_feature(enable = "zbkb")]
+#[cfg_attr(test, assert_instr(brev8))]
+#[inline]
+pub unsafe fn brev8(rs: usize) -> usize {
+    let value: usize;
+    unsafe {
+        asm!(
+            "brev8 {rd},{rs}",
+            rd = lateout(reg) value,
+            rs = in(reg) rs,
+            options(pure, nomem, nostack),
+        )
+    }
+    value
+}
+
+/// Byte-wise lookup of indicies into a vector in registers.
+///
+/// The xperm8 instruction operates on bytes. The rs1 register contains a vector of XLEN/8
+/// 8-bit elements. The rs2 register contains a vector of XLEN/8 8-bit indexes. The result is
+/// each element in rs2 replaced by the indexed element in rs1, or zero if the index into rs2
+/// is out of bounds.
+///
+/// Source: RISC-V Cryptography Extensions Volume I: Scalar & Entropy Source Instructions
+///
+/// Version: v1.0.1
+///
+/// Section: 3.47
+///
+/// # Safety
+///
+/// This function is safe to use if the `zbkx` target feature is present.
+#[target_feature(enable = "zbkx")]
+#[cfg_attr(test, assert_instr(xperm8))]
+#[inline]
+pub unsafe fn xperm8(rs1: usize, rs2: usize) -> usize {
+    let value: usize;
+    unsafe {
+        asm!(
+            "xperm8 {rd},{rs1},{rs2}",
```

Comment:
Prefer using LLVM intrinsics over inline asm. For example, in this case you can just call `llvm.riscv.xperm8.i64`. Refer to `llvm-project/clang/test/CodeGen/RISCV/rvb-intrinsics/` for the exact names of the LLVM intrinsics to use.